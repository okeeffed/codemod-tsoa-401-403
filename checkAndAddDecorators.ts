import {
  Project,
  ClassDeclaration,
  MethodDeclaration,
  Decorator,
} from "ts-morph";

export function checkAndAddDecorators(filePaths: string[]): {
  [filePath: string]: string;
} {
  const project = new Project();
  project.addSourceFilesAtPaths(filePaths);

  const sourceFiles = project.getSourceFiles();
  const modifiedFiles: { [filePath: string]: string } = {};

  sourceFiles.forEach((sourceFile) => {
    let fileModified = false;
    const classes = sourceFile.getClasses();

    classes.forEach((classDeclaration: ClassDeclaration) => {
      const methods = classDeclaration.getMethods();

      methods.forEach((method: MethodDeclaration) => {
        const decorators = method.getDecorators();
        const hasUnauthorized = hasResponseDecorator(
          decorators,
          401,
          "Unauthorized"
        );
        const hasForbidden = hasResponseDecorator(decorators, 403, "Forbidden");

        if (!hasUnauthorized) {
          addResponseDecorator(method, 401, "Unauthorized");
          fileModified = true;
        }
        if (!hasForbidden) {
          addResponseDecorator(method, 403, "Forbidden");
          fileModified = true;
        }
      });
    });

    if (fileModified) {
      modifiedFiles[sourceFile.getFilePath()] = sourceFile.getFullText();
    }
  });

  return modifiedFiles;
}

function hasResponseDecorator(
  decorators: Decorator[],
  statusCode: number,
  description: string
): boolean {
  return decorators.some((decorator) => {
    const callExpression = decorator.getCallExpression();
    if (
      callExpression &&
      callExpression.getExpression().getText() === "Response"
    ) {
      const args = callExpression.getArguments();
      return (
        args.length >= 2 &&
        args[0].getText().includes(statusCode.toString()) &&
        args[1].getText().includes(description)
      );
    }
    return false;
  });
}

function addResponseDecorator(
  method: MethodDeclaration,
  statusCode: number,
  description: string
) {
  method.addDecorator({
    name: "Response",
    arguments: [`${statusCode}, "${description}"`],
    typeArguments: ["ErrorResponse"],
  });
  console.log(
    `Added @Response<ErrorResponse>(${statusCode}, "${description}") to method '${method.getName()}'`
  );
}

// Get file paths from command-line arguments
// TODO: Mod this to search for *.controller.ts files
const filesToCheck = process.argv.slice(2);
if (filesToCheck.length === 0) {
  console.log(
    "No files specified. Usage: ts-node checkAndAddDecorators.ts <file1> <file2> ..."
  );
} else {
  checkAndAddDecorators(filesToCheck);
}
