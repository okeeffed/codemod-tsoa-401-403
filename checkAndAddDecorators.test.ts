import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { checkAndAddDecorators } from "./checkAndAddDecorators";
import fs from "fs";
import path from "path";

describe("checkAndAddDecorators", () => {
  const testFilePath = path.resolve(__dirname, "__tests__/test.controller.ts");
  const backupFilePath = `${testFilePath}.backup`;

  beforeEach(() => {
    // Create a backup of the original file
    fs.copyFileSync(testFilePath, backupFilePath);
  });

  afterEach(() => {
    // Restore the original file from backup
    fs.copyFileSync(backupFilePath, testFilePath);
    fs.unlinkSync(backupFilePath);
  });

  it("should add missing decorators", () => {
    const originalContent = fs.readFileSync(testFilePath, "utf8");
    console.log("Original content:", originalContent);

    const modifiedFiles = checkAndAddDecorators([testFilePath]);
    const updatedContent = modifiedFiles[testFilePath] || originalContent;
    console.log("Updated content:", updatedContent);

    const unauthorizedRegex =
      /@Response<ErrorResponse>\(401, "Unauthorized"\)/g;
    const forbiddenRegex = /@Response<ErrorResponse>\(403, "Forbidden"\)/g;

    expect(unauthorizedRegex.test(updatedContent)).toBe(true);
    expect(forbiddenRegex.test(updatedContent)).toBe(true);
    expect(updatedContent).not.toBe(originalContent);
  });

  it("should not add decorators if they already exist", () => {
    // First, add the decorators
    const modifiedFiles1 = checkAndAddDecorators([testFilePath]);
    const contentWithDecorators = modifiedFiles1[testFilePath] || "";

    // Then, try to add them again
    const modifiedFiles2 = checkAndAddDecorators([testFilePath]);
    const finalContent = modifiedFiles2[testFilePath] || contentWithDecorators;

    expect(finalContent).toBe(contentWithDecorators);
  });
});
