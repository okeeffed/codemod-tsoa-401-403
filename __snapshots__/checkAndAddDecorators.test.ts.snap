// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkAndAddDecorators > should not add decorators if they already exist 1`] = `
"
import { Response, ErrorResponse } from 'some-library';

class TestController {
  @SomeDecorator()
  @Response<ErrorResponse>(401, "Unauthorized")
  @Response<ErrorResponse>(403, "Forbidden")
  public testMethod() {
    // Method implementation
  }
}
    "
`;
